### ISSUE TITLE
Automatic API for postproc functions does not work

### ISSUE THREAD

**Issue by imaliyov** [perturbo-code/perturbopy#92]
https://perturbopy.readthedocs.io/en/latest/_autosummary/perturbopy.html#module-perturbopy

Here, the postproc API does not work. One cannot click on it. Everything else there works, but postproc is the mostimportant. 
One has to check the generation with sphynx.

This can be done after the release but, still urgently.

**Comment by ltan01** [comment]
Hi Ivan, I will look into this.

**Comment by ltan01** [comment]
Hi Ivan @imaliyov, when I generate the website locally, all of the API buttons work. There were no differences between my local branch and main.

I'm wondering if the gitignore _autosummary file is related to this issue? Why are we ignoring this?

**Comment by imaliyov** [comment]
I think the idea is that the autosummary should be generated on the server. Otherwise, you would spam you commits.

It means that readthedocs does not work correctly

**Comment by ltan01** [comment]
We tried troubleshooting it, but we'll look at it after the release.

**Comment by imaliyov** [comment]
Sounds good. Thank you Laurie!
I actually wanted to reference to Perturbopy function parameters from my spectra tutorial and could not, that's how I mentioned :) 
It could be really cool to make it work.

**Comment by imaliyov** [comment]
Here is `make html` in the docs folder:
<img width="870" alt="image" src="https://github.com/user-attachments/assets/47fcccfd-72a7-4621-a303-15464091631f" />

It does give warnings related to the postproc module. I think this is the reason.
Do you get similar warnings on your end?

**Comment by hurricane642** [comment]
> Here is `make html` in the docs folder: <img alt="image" width="870" src="https://private-user-images.githubusercontent.com/24958411/396412457-47fcccfd-72a7-4621-a303-15464091631f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ0NTAyNzQsIm5iZiI6MTczNDQ0OTk3NCwicGF0aCI6Ii8yNDk1ODQxMS8zOTY0MTI0NTctNDdmY2NjZmQtNzJhNy00NjIxLWEzMDMtMTU0NjQwOTE2MzFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE3VDE1MzkzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQxZGQwNzhhODI4YjI1YTMyY2MwYTgxYmE3ZjdlNTI2YmFmY2M2ODZkZmU5NzlmNjNhMWQ2NTk3YWJiMzFiNTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SgFqTzqx58BUNK8iunHmM_WmzbZairk71KpLD57Oxys">
> 
> It does give warnings related to the postproc module. I think this is the reason. Do you get similar warnings on your end?

Yep, I have the same issue. Maybe, we need to check the `postproc` module accurately

**Comment by imaliyov** [comment]
Something should call the functions in postproc or import it for readthedocs to work, I guess...

**Comment by imaliyov** [comment]
If the functions are not called or imported somewhere, this might cause Sphynx to not 'find' them.

**Comment by hurricane642** [comment]
Hey @imaliyov ,
Me and @ltan01 found the issue - it is related to the `__init__.py` in the `postproc` folder. If you remover all of this complicated imports, API works exactly as it should. But, without this imports, usage of perturbopy become much more ugly...
Do you have any ideas or suggestions on what we can do here?

**Comment by imaliyov** [comment]
Could you please show here an __init__ that we have currently on master branch of perturbopy
and __initi__ that works with the API?

In general, I think, we are doing a lot of weird things when importing different modules within the package itself :) 
Would be nice to see how this is handled in other good Python packages.

Maybe relative paths are ok within the package?..

**Comment by hurricane642** [comment]
> Could you please show here an **init** that we have currently on master branch of perturbopy and **initi** that works with the API?
> 
> In general, I think, we are doing a lot of weird things when importing different modules within the package itself :) Would be nice to see how this is handled in other good Python packages.
> 
> Maybe relative paths are ok within the package?..

Current `__init__.py`:
```
"""
Suite of Python scripts for the Perturbo code testing and postprocessing.
"""

from .calc_modes.calc_mode import CalcMode
from .calc_modes.bands import Bands
from .calc_modes.spectral_cumulant import SpectralCumulant
from .calc_modes.spins import Spins
from .calc_modes.phdisp import Phdisp
from .calc_modes.ephmat import Ephmat
from .calc_modes.ephmat_spin import EphmatSpin
from .calc_modes.trans import Trans
from .calc_modes.imsigma import Imsigma
from .calc_modes.imsigma_spin import ImsigmaSpin
from .calc_modes.dyna_run import DynaRun
from .calc_modes.dyna_pp import DynaPP

from .dbs.units_dict import UnitsDict
from .dbs.recip_pt_db import RecipPtDB

from .utils import constants, plot_tools, lattice, spectra_generate_pulse, \
    timing, spectra_trans_abs, spectra_plots

import warnings
import os


# Define a custom format for warnings
def custom_formatwarning(message, category, filename, lineno, line=None):
    basename = os.path.basename(filename)
    return f"{category.__name__} in {basename} at line {lineno}: {message}\n"


# Apply the custom format
warnings.formatwarning = custom_formatwarning

```
`__init__.py` which generates documentation in the right way:
```

"""
Suite of Python scripts for the Perturbo code testing and postprocessing.
"""

```
I'll try to check other packages, I think, that we simplify usage of package by all of this imports

**Comment by imaliyov** [comment]
Yeah, this one looks alright I think...

I suggest to find a hacky way to tell the Sphynx API to use another __init__ or something like this.

We don't want to significantly change the structure of the package to just make Sphynx work.
On the other hand, having the API docs is very nice, so some compromise could be best! Don't know how much this helps :)

**Comment by hurricane642** [comment]
Hey @imaliyov , 
We've continued work on the API with @ltan01 , this is our results:
1. We've tried to install and run readthedocs locally, but documentation is embarrassingly outdated and perplexed, so it was not the option.
2. During the documentation building, we've met this error: `WARNING: duplicate object description of ...` which was related exactly to postproc. We think the problem is really that each class in `postproc` is contained twice - in the corresponding folder and in the import in `__init__.py`. In the case of a local run, this creates just a warning, but generates the documentation as it should. Apparently in the case of readthedocs this creates an error.
3. Seems like it's a popular bug of the readthedocs: https://github.com/readthedocs/sphinx-autoapi/issues/358 , https://github.com/readthedocs/sphinx-autoapi/issues/317#issuecomment-1033316693 etc. 

From this, we have following suggestion. If the package is organized in the right way and sphinx itself work as it should, maybe we need to switch from readthedocs to something else? Actually, seems like the easiest alternative is to use Github Pages, We'll need to update reference links, but it could be solution in this case.
What do you think?

**Comment by imaliyov** [comment]
I would not go this way, honestly.

Readthedocs has lots of configurations that are different with GitHub pages. I know that because the Perturbo-code website is, as you know, hosted on GitHub pages :) 

Worst case scenario, I would forget about the API problem, but keep the code on readthedocs.. just an idea to get my feeling about this.

> We think the problem is really that each class in postproc is contained twice - in the corresponding folder and in the import in __init__.py.

Now, this makes sense! This behavior would trick the API in term of dependencies, references, etc.
Don't you think that this is bad overall, setting readthedocs aside? Anyways, it generates a warning for a reason, so even if the code runs, it is not the best thing to do on the first place.
I'd suggest to fix this thing, I promise, it is simpler than migrate the website to another host! :)

**Comment by hurricane642** [comment]
> I would not go this way, honestly.
> 
> Readthedocs has lots of configurations that are different with GitHub pages. I know that because the Perturbo-code website is, as you know, hosted on GitHub pages :)
> 
> Worst case scenario, I would forget about the API problem, but keep the code on readthedocs.. just an idea to get my feeling about this.
> 
> > We think the problem is really that each class in postproc is contained twice - in the corresponding folder and in the import in **init**.py.
> 
> Now, this makes sense! This behavior would trick the API in term of dependencies, references, etc. Don't you think that this is bad overall, setting readthedocs aside? Anyways, it generates a warning for a reason, so even if the code runs, it is not the best thing to do on the first place. I'd suggest to fix this thing, I promise, it is simpler than migrate the website to another host! :)

Make sense:)
I think, that general approach of import in `__init__.py` file is right, it significantly simplify further usage. But maybe we can try to check the same idea is done in different packages.
I agree that migration to different host is a bit overkill for this problem ahah, it is not so significant

**Comment by imaliyov** [comment]
Nice! So, ideally, we'd try to fix the __init__.py files, to remove that warning, and then perhaps, readthedocs will start working too.

**Comment by hurricane642** [comment]
Solved
